File: packages/provide-uterm-csharp/src/Provide.Uterm/Colors/Sgr.cs
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

using System.Text.RegularExpressions;

namespace Provide.Uterm.Colors;

/// <summary>SGR parameter-list rewriting for color downgrade.</summary>
public static partial class Sgr
{
    public static readonly Regex SgrRegexp = SgrPattern();

    [GeneratedRegex(@"\x1b\[([0-9;]*)m")]
    private static partial Regex SgrPattern();

    private static readonly int[] Fg16 = [30, 34, 32, 36, 31, 35, 33, 37, 90, 94, 92, 96, 91, 95, 93, 97];
    private static readonly int[] Bg16 = [40, 44, 42, 46, 41, 45, 43, 47, 100, 104, 102, 106, 101, 105, 103, 107];

    private static bool IsDigits(string s)
    {
        if (s.Length == 0)
        {
            return false;
        }

        foreach (var c in s)
        {
            if (c is < '0' or > '9')
            {
                return false;
            }
        }

        return true;
    }

    private static int ParseComponent(string s)
    {
        var trimmed = s.TrimStart('0');
        if (trimmed.Length > 9)
        {
            return 1 << 30;
        }

        return int.TryParse(s, out var v) ? v : 0;
    }

    public static string RewriteParams(string parameters, ColorMode mode)
    {
        if (parameters.Length == 0)
        {
            return "\x1b[" + parameters + "m";
        }

        var parts = parameters.Split(';');
        var output = new List<string>(parts.Length);
        var i = 0;
        var n = parts.Length;
        while (i < n)
        {
            if (i + 4 < n &&
                (parts[i] == "38" || parts[i] == "48") &&
                parts[i + 1] == "2" &&
                IsDigits(parts[i + 2]) && IsDigits(parts[i + 3]) && IsDigits(parts[i + 4]))
            {
                var r = ParseComponent(parts[i + 2]);
                var g = ParseComponent(parts[i + 3]);
                var b = ParseComponent(parts[i + 4]);
                var isFg = parts[i] == "38";
                if (mode == ColorMode.Mode256)
                {
                    var code = Rgb.RgbTo256(r, g, b);
                    output.Add(isFg ? "38" : "48");
                    output.Add("5");
                    output.Add(code.ToString());
                }
                else
                {
                    var idx = Rgb.RgbTo16Index(r, g, b);
                    output.Add((isFg ? Fg16[idx] : Bg16[idx]).ToString());
                }

                i += 5;
                continue;
            }

            output.Add(parts[i]);
            i++;
        }

        return "\x1b[" + string.Join(";", output) + "m";
    }
}

File: packages/provide-uterm-ts/src/colors/sgr.ts
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

/**
 * SGR parameter-list rewriting for color downgrade.
 *
 * Given an SGR parameter list (the `N;M;...` part between `ESC [` and `m`),
 * scan for truecolor runs (`38;2;R;G;B` foreground or `48;2;R;G;B`
 * background) and replace them with the configured lower-palette
 * equivalent. Other SGR parameters (bold, italic, 256-color, etc.) pass
 * through unchanged.
 *
 * Port of the Python module `provide.uterm.colors.sgr`.
 */

import { rgbTo16Index, rgbTo256 } from "./rgb.ts";

/**
 * Target palette for {@link rewriteParams}. `"passthrough"` is not a valid
 * rewrite target — the dispatcher handles it before reaching here.
 */
export type DowngradeTarget = "256" | "16";

/**
 * Build the SGR escape-sequence pattern: `\x1b[` params `m`, where params is
 * a possibly-empty semicolon-separated parameter list.
 *
 * Returned fresh per call because a `g`-flagged RegExp carries mutable
 * `lastIndex` state that would leak between scans.
 */
export function sgrPattern(): RegExp {
  return /\x1b\[([0-9;]*)m/g;
}

/**
 * Base ANSI 16-color foreground escape codes, indexed by palette index 0-15
 * (0-7 are the normal colors, 8-15 the bright variants).
 */
const FG_16 = [30, 34, 32, 36, 31, 35, 33, 37, 90, 94, 92, 96, 91, 95, 93, 97];
/** Base ANSI 16-color background escape codes, indexed by palette index 0-15. */
const BG_16 = [40, 44, 42, 46, 41, 45, 43, 47, 100, 104, 102, 106, 101, 105, 103, 107];

/**
 * Report whether `s` is a non-empty run of ASCII decimal digits.
 *
 * Stands in for Python's `str.isdigit()`. The SGR pattern only ever yields
 * `[0-9;]`, so within the scanner the two agree exactly; on a direct call
 * with a non-digit the run is passed through instead of raising.
 */
function isDigits(s: string): boolean {
  if (s === "") {
    return false;
  }
  for (let i = 0; i < s.length; i += 1) {
    const c = s.charCodeAt(i);
    if (c < 0x30 || c > 0x39) {
      return false;
    }
  }
  return true;
}

/**
 * Rewrite an SGR parameter list, downgrading any truecolor runs.
 *
 * Walks the `;`-separated parameters looking for the 5-parameter run
 * `38;2;R;G;B` (foreground truecolor) or `48;2;R;G;B` (background
 * truecolor). Each such run is replaced with its equivalent under the target
 * mode; everything else is preserved in place and order.
 *
 * @param params SGR parameter list without the leading `\x1b[` or trailing
 *   `m`. May be empty.
 * @param mode `"256"` to map to the xterm-256 cube, `"16"` to map to the
 *   base 16-color palette.
 * @returns A full SGR escape sequence (`\x1b[<rewritten>m`).
 */
export function rewriteParams(params: string, mode: DowngradeTarget): string {
  if (params === "") {
    return `\x1b[${params}m`;
  }
  const parts = params.split(";");
  const out: string[] = [];
  const n = parts.length;
  let i = 0;
  while (i < n) {
    const head = parts[i] as string;
    if (
      i + 4 < n &&
      (head === "38" || head === "48") &&
      parts[i + 1] === "2" &&
      isDigits(parts[i + 2] as string) &&
      isDigits(parts[i + 3] as string) &&
      isDigits(parts[i + 4] as string)
    ) {
      const r = Number(parts[i + 2]);
      const g = Number(parts[i + 3]);
      const b = Number(parts[i + 4]);
      const isFg = head === "38";
      if (mode === "256") {
        out.push(isFg ? "38" : "48", "5", String(rgbTo256(r, g, b)));
      } else {
        const idx = rgbTo16Index(r, g, b);
        out.push(String((isFg ? FG_16[idx] : BG_16[idx]) as number));
      }
      i += 5;
      continue;
    }
    out.push(head);
    i += 1;
  }
  return `\x1b[${out.join(";")}m`;
}
